### PR TITLE
Customizable `enforceContactAndJointAndCustomConstraints` function

### PR DESCRIPTION
**Pull request by michguo**
Round 3 Review
- Fix unittests and python code still using `world`
- Give a more descriptive name to custom function (`solveCallback` -> `enforceContactAndJointAndCustomConstraints`)

Round 2 Review
- Remove dependency on `world` inputs.
- Add warning message about not supporting differentiation through custom callback functions.

Round 1 Review
- Create `mSolveCallback` callback function.

### CODE REVIEW

File: python/new_examples/constraint_callback.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    action = torch.zeros((world.getNumDofs()))
+    solver = world.getConstraintSolver()
+    callbacks = [None, dummy_callback, solver.lcpSolveCallback, lcp_callback, frictionless_lcp_callback]
+    for callback in callbacks:
+        if callback is not None:
+            solver.replaceSolveCallback(callback)
+            print(callback.__name__)
+        else:
+            print("None")
+        new_state = nimble.timestep(world, state, action)
```
Comment (keenon) [inline python/new_examples/constraint_callback.py:41]: I worry that this will no longer do the right thing when we try to differentiate through it. Because the current backwards pass assumes it's always an LCP, it relies on lines in `BoxedLcpConstraintSolver.cpp` like:

```
group.getGradientConstraintMatrices()->registerLCPResults(
        mX,
        hiGradientBackup,
        loGradientBackup,
        fIndexGradientBackup,
        bGradientBackup,
        aColNormGradientBackup,
        aGradientBackup,
        cfm,
        hadToIgnoreFrictionToSolve);
    group.getGradientConstraintMatrices()->constructMatrices(world);
```
Then those constrained groups get used during the computation of `ConstrainedGroupGradientMatrices`'s various Jacobians.

The problem is, if we're using a solver callback, none of that machinery will work.

I think for now, it'd be fine to just flash a big prominent warning if you try to backwards pass, but you've got a `replaceSolveCallback()` in place, like "WARNING: GRADIENTS WILL BE INCORRECT!!!! Nimble is still under heavy development, and we don't yet support differentiating through `timestep()` if you've called `replaceSolveCallback()` to customize the contact model." or something...

File: python/_nimblephysics/simulation/World.cpp [hunk 2]
```diff
@@ -169,6 +169,17 @@ void World(py::module& m)
             return self->getIndex(_index);
           },
           ::py::arg("index"))
+      .def(
+          "getNumBodyNodes",
+          +[](dart::simulation::World* self) -> std::size_t {
+            return self->getNumBodyNodes();
+          })
+      .def(
+          "getBodyNodeIndex",
```
Comment (keenon) [inline python/_nimblephysics/simulation/World.cpp:178]: I see you didn't create this method, but is it really the case that `getBodyNodeIndex()` doesn't return the index of a BodyNode? I think we should rename this to explain better what it does. Maybe `getBodyNodeByIndex()`?

**Review by keenon — COMMENTED** [review]
Ok, first of all, this is insanely cool. I'm so excited to see this coming together so fast!

I think we have to spend some time thinking about eventual differentiation, and making sure we plan for that. We'll probably want to merge this in before the differentiation is completely done, so let's make sure we add prominent warnings where things are going to unexpectedly break for users.

File: python/_nimblephysics/simulation/World.cpp (hunk 2 above)
Comment (michguo) [reply python/_nimblephysics/simulation/World.cpp:178]: Agreed! Opened #59 to fix this in a separate PR.

File: python/_nimblephysics/constraint/ConstraintSolver.cpp [hunk 3]
```diff
@@ -200,7 +201,16 @@ void ConstraintSolver(py::module& m)
       .def(
           "solve",
           +[](dart::constraint::ConstraintSolver* self,
-              dart::simulation::World* world) { self->solve(world); });
+              dart::simulation::World* world) { self->solve(world); })
+      .def(
+          "lcpSolveCallback",
+          +[](dart::constraint::ConstraintSolver* self,
+              dart::simulation::World* world) {
```
Comment (keenon) [inline python/_nimblephysics/constraint/ConstraintSolver.cpp:208]: I suspect you just forgot to change this signature. You don't need the `world` argument anymore, no?

File: python/new_examples/constraint_callback.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+"""An example of overriding the ConstraintSolver's default callback function."""
+import nimblephysics as nimble
+import torch
+
+
+def dummy_callback(world: nimble.simulation.World):
+    pass
+
+def lcp_callback(world: nimble.simulation.World):
+    world.getConstraintSolver().lcpSolveCallback(world)
```
Comment (keenon) [inline python/new_examples/constraint_callback.py:10]: You should remove `world` here also, no?

**Review by keenon — COMMENTED** [review]
I know you're still working on this, so I'll just leave these comments here

**Review by keenon — APPROVED** [review]
